pkg/pillar/devicenetwork/dhcpcd.go

```diff
@@ -211,16 +211,44 @@ func doDhcpClientInactivate(log *base.LogObject, nuc types.NetworkPortConfig) {
 		log.Functionf("doDhcpClientInactivate(%s) DT_NONE is a no-op\n",
 			nuc.IfName)
 	case types.DT_STATIC, types.DT_CLIENT:
-		extras := []string{}
-		if !dhcpcdCmd(log, "--release", extras, nuc.IfName, false) {
-			log.Errorf("doDhcpClientInactivate: release failed for %s\n",
-				nuc.IfName)
+		var extras []string
+		// Wait for a bit then give up
+		waitCount := 0
+		failed := false
+		for {
+			//it waits up to 10 seconds for exit https://github.com/NetworkConfiguration/dhcpcd/blob/dhcpcd-8.1.6/src/dhcpcd.c#L1950-L1957
+			if !dhcpcdCmd(log, "--release", extras, nuc.IfName, false) {
+				log.Errorf("doDhcpClientInactivate: release failed for %s", nuc.IfName)
+			}
+			if !dhcpcdExists(log, nuc.IfName) {
+				break
+			}
+			waitCount++
+			if waitCount >= 3 {
+				log.Errorf("doDhcpClientInactivate: dhcpcd %s still running, will kill it", nuc.IfName)
+				failed = true
+				break
+			}
+			log.Warnf("doDhcpClientInactivate: dhcpcd %s still running", nuc.IfName)
+			time.Sleep(1 * time.Second)
```

Comment:
You're only waiting for 3 seconds here and dhcpcd has a 10 second timeout you said.
Makes sense to wait for 10 seconds before giving up.

Reply:
We run it in foreground and wait for ends

Reply:
Are you saying that dhcpcd --release is waiting for the release to complete and the process exit?

The code isn't assuming that but is assuming that it needs to check for the process to be gone.

Reply:
It wait for exit or time out. But the code actually do the correct thing, it runs the command and check current status of process.

Reply:
Sure, but what is the shortest and longest time we can spend in this loop?
One way to read it is that we can give up after 3 seconds (which is too short) if dhcpcd --release just sends a signal to the running dhcpcd process.
But if dhcpcd --release waits for the running dhcpcd process to release the lease with the server (with some timeout), then this loop can take a lot longer.

Reply:
It waits up to 10 seconds (as I mentioned inside comment with link to code), so the longest time here is about 32 seconds.